### ISSUE TITLE
Fix flaky ReentrantLockSpec - Cleans up interrupted waiters

### ISSUE THREAD

**Issue by justcoon** [zio/zio#6520]
flaky ReentrantLockSpec

```
[info] - ReentrantLockSpec
[info]   + 1 lock
[info]   + 2 locks
[info]   + 2 locks from different fibers
[info]   - Cleans up interrupted waiters
[info]     `waiters1` = 1 did not satisfy equalTo(2)
[info]     at /home/runner/work/zio/zio/concurrent/shared/src/test/scala/zio/concurrent/ReentrantLockSpec.scala:53
[info]   + Fairness assigns lock to fibers in order
[info]   + Assigns lock to fibers randomly
[info] Ran 6 tests in 1 s 467 ms: 5 succeeded, 0 ignored, 1 failed
[info] - ReentrantLockSpec
[info]   - Cleans up interrupted waiters
[info]     `waiters1` = 1 did not satisfy equalTo(2)
[info]     at /home/runner/work/zio/zio/concurrent/shared/src/test/scala/zio/concurrent/ReentrantLockSpec.scala:53
[info] Done
[error] Failed tests:
[error] 	zio.concurrent.ReentrantLockSpec
[error] (concurrentJVM / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 9 s, completed Mar 23, 2022, 3:00:35 PM
```

https://github.com/zio/zio/runs/5662383425?check_suite_focus=true

**Comment by adamgfraser** [comment]
Thanks for reporting! I have also observed this.

**Comment by adamgfraser** [comment]
Copying @regiskuckaertz.

**Comment by regiskuckaertz** [comment]
Ooh, ok I will take a look, thanks for reporting!

**Comment by regiskuckaertz** [comment]
Closed by #6523
